test/cri-containerd/container_layers_packing_test.go

```diff
@@ -0,0 +1,171 @@
+// +build functional
+
+package cri_containerd
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/Microsoft/hcsshim/internal/shimdiag"
+	"github.com/Microsoft/hcsshim/osversion"
+	testutilities "github.com/Microsoft/hcsshim/test/functional/utilities"
+)
+
+const (
+	ubuntu1804          = "ubuntu:18.04"
+	ubuntu70ExtraLayers = "cplatpublic.azurecr.io/ubuntu70extra:18.04"
+	alpine70ExtraLayers = "cplatpublic.azurecr.io/alpine70extra:latest"
+)
+
+func filterStrings(input []string, include string) []string {
+	var result []string
+	for _, str := range input {
+		if strings.Contains(str, include) {
+			result = append(result, str)
+		}
+	}
+	return result
+}
+
+func shimDiagExec(ctx context.Context, t *testing.T, podID string, cmd []string) string {
+	shimName := fmt.Sprintf("k8s.io-%s", podID)
+	shim, err := shimdiag.GetShim(shimName)
+	if err != nil {
+		t.Fatalf("failed to find shim %v: %v", shimName, err)
+	}
+	shimClient := shimdiag.NewShimDiagClient(shim)
+
+	bufOut := &bytes.Buffer{}
+	bw := bufio.NewWriter(bufOut)
+	bufErr := &bytes.Buffer{}
+	bwErr := bufio.NewWriter(bufErr)
+
+	exitCode, err := execInHost(ctx, shimClient, cmd, nil, bw, bwErr)
+	if err != nil {
+		t.Fatalf("failed to exec request in the host with: %v and %v", err, bufErr.String())
+	}
+	if exitCode != 0 {
+		t.Fatalf("exec request in host failed with exit code %v: %v", exitCode, bufErr.String())
+	}
+
+	return strings.TrimSpace(bufOut.String())
+}
+
+func validateTargets(ctx context.Context, t *testing.T, podID string, expected int) {
+	dmDiag := shimDiagExec(ctx, t, podID, []string{"ls", "-l", "/dev/mapper"})
+	dmLines := filterStrings(strings.Split(dmDiag, "\n"), "dm-linear")
+
+	lrDiag := shimDiagExec(ctx, t, podID, []string{"ls", "-l", "/run/layers"})
+	lrLines := filterStrings(strings.Split(lrDiag, "\n"), "p")
+	if len(lrLines) != len(dmLines) {
+		t.Log(dmDiag)
+		t.Log(lrDiag)
+		t.Fatalf("number of layers and device-mapper targets mismatch")
```

Comment:
nit: you can include dmDiag and lrDiag in the fatalf call instead of logging them separately